op-program/host/kvstore/kv_test.go

```diff
@@ -0,0 +1,67 @@
+package kvstore
+
+import (
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/stretchr/testify/require"
+)
+
+func kvTest(t *testing.T, kv KV) {
+	t.Run("roundtrip", func(t *testing.T) {
+		t.Parallel()
+		_, err := kv.Get(common.Hash{0xaa})
+		require.Equal(t, err, ErrNotFound, "file (in new tmp dir) does not exist yet")
+
+		require.NoError(t, kv.Put(common.Hash{0xaa}, []byte("hello world")))
+		dat, err := kv.Get(common.Hash{0xaa})
+		require.NoError(t, err, "pre-image must exist now")
+		require.Equal(t, "hello world", string(dat), "pre-image must match")
+	})
+
+	t.Run("empty pre-image", func(t *testing.T) {
+		t.Parallel()
+		require.NoError(t, kv.Put(common.Hash{0xbb}, []byte{}))
+		dat, err := kv.Get(common.Hash{0xbb})
+		require.NoError(t, err, "pre-image must exist now")
+		require.Zero(t, len(dat), "pre-image must be empty")
+	})
+
+	t.Run("zero pre-image key", func(t *testing.T) {
+		t.Parallel()
+		// in case we give a pre-image a special empty key. If it was a hash then we wouldn't know the pre-image.
+		require.NoError(t, kv.Put(common.Hash{}, []byte("hello")))
+		dat, err := kv.Get(common.Hash{})
+		require.NoError(t, err, "pre-image must exist now")
+		require.Equal(t, "hello", string(dat), "pre-image must match")
+	})
+
+	t.Run("non-string value", func(t *testing.T) {
+		t.Parallel()
+		// in case we give a pre-image a special empty key. If it was a hash then we wouldn't know the pre-image.
+		require.NoError(t, kv.Put(common.Hash{0xcc}, []byte{4, 2}))
+		dat, err := kv.Get(common.Hash{0xcc})
+		require.NoError(t, err, "pre-image must exist now")
+		require.Equal(t, []byte{4, 2}, dat, "pre-image must match")
+	})
+
+	t.Run("not overwriting pre-image", func(t *testing.T) {
+		t.Parallel()
+		require.NoError(t, kv.Put(common.Hash{0xdd}, []byte{4, 2}))
+		require.ErrorIs(t, kv.Put(common.Hash{0xdd}, []byte{4, 2}), ErrAlreadyExists)
+	})
+
+	t.Run("concurrent", func(t *testing.T) {
+		hash := common.Hash{0xee}
+		data := []byte{1, 2, 3, 4, 5}
+		go func() {
+			err := kv.Put(hash, data)
+			require.NoError(t, err)
+		}()
+		time.Sleep(3 * time.Second)
```

Comment:
What's the purpose of this test? And why does it have to have such a high sleep time? Wouldn't something like 100 ms be sufficient? Or even something like
```suggestion
        done := make(chan(struct{}))
		go func() {
			err := kv.Put(hash, data)
			require.NoError(t, err)
			close(done)
		}()
		<-done
```
to avoid sleeping in the test? Although I don't see what the test gives us.

In general I'd try to avoid long sleeps in tests if possible.

If the purpose is to test concurrent access, I'd spawn like 100 goroutines and run the roundtrip test in each, so that concurrent read&write access is tested.

Reply:
Hmm, the proper test of the get-put is only really reproducible by starting the put, then doing the get, and then ending the put. But that can't work with rwlock. And by closing the chan after Put, the Get is already synchronized to happen after put completes. To test we'd have to mock a slow file write, which requires more complexity (have to mock a read-write filesystem (go has read-only FS abstraction, but nothing nice for writes), and it wouldn't be a generic KV test anymore either. I'm dropping the test.